### ISSUE TITLE
ER: $x.a = 0

### ISSUE THREAD

**Issue by pkoppstein** [stedolan/jq#1610]
Consider:

```
$ jq -nc '{} as $x | ($x |.a = 0)'
{"a":0}
```

but the considerably more readable:

```
jq -nc '{} as $x | $x.a = 0'
```
results in:

    jq: error (at <unknown>): Invalid path expression near attempt to access element "a" of {}

**Comment by nicowilliams** [comment]
This error occurs because the `$x` happens in the `paths` closure here:

```
def _assign(paths; value): value as $v | reduce path(paths) as $p (.; setpath($p; $v));
```

which implements assignment.

Well, a `LOADV` is necessarily not permitted in `path/1`'s argument expression, but that's what we're doing here!  There's no way to fix this other than by recognizing the pattern and moving the `$x` / `LOADV` outside the `paths` expression, outside the assignment.

I.e., the compiler will have to recognize `($x.a = ...)` and rewrite as `($x | (.a = ...))`.

Note that `($x.a, $y.b) = ...` cannot be permitted, so the compiler would have to distinguish the two cases.  As it happens, that's not hard at this time, but still...

All in all, fixing this will either be non-trivial or not-pretty.

**Comment by nicowilliams** [comment]
To expand a bit, `... | (($x | .a) = ...)` is necessarily an error, but we can rewrite it as `... | ($x | (.a = ...))` when we can tell that that is a safe transformation.  Can we trivially tell that's a safe transformation?  Actually, I'm not sure.  Consider this:

```
$ jq -cn '. as $x | inputs | $x.a = 0'
```

What is the intention there?  Is it to consume all inputs and for each output `{"a":0}`, or is it some error on the part of the programmer??

I'm considering closing this as wontfix.  Advice?

**Comment by pkoppstein** [comment]
> Advice?

My *hope* is that 1.6 will take priority, and that this issue be considered only after all the much more pressing issues are resolved. If the mere fact that this Issue is "open" is distracting, then by all means close it.  

I should have noted that the manual does explicitly mention that `$var.foo = 1` is unsupported, but one can reasonably take the view that the whole point of the magic behind assignment in jq is to make things easy in an intuitive way, so if some voodoo is necessary to make it work for $-variables, it probably won't be the first time.

**Comment by nicowilliams** [comment]
Fair enough.

**Comment by pkoppstein** [comment]
@nicowilliams - I changed the title so it's clear this is an ER.  Feel free to rm the `bug` tag :-)

**Comment by nicowilliams** [comment]
Actually, considering that the manual already says this is not supported, and that trying to make it work has ambiguity problems, the better path forward is to have this raise a compile-time error -- much more useful.
